src/main/java/algorithms/medium/CheapestFlightsWithinKStops.java

```diff
@@ -0,0 +1,84 @@
+package algorithms.medium;
+
+import java.util.PriorityQueue;
+
+public class CheapestFlightsWithinKStops {
+
+    public int findCheapestPrice(int n, int[][] flights, int src, int dst, final int K) {
+        int[][][] graph = createGraph(flights, n);
+        return dijkstra(graph, src, dst, K, n);
+    }
+
+    int[][][] createGraph(int[][] arcs, int n) {
+        int[][][] graph = new int[n][][];
+
+        int[] outDegree = new int[n + 1];
+        for (int[] arc : arcs) {
+            outDegree[arc[0]]++;
+        }
+
+        for (int i = 0; i < n; i++) {
+            graph[i] = new int[outDegree[i]][2];
+        }
+        int[] currIdx = new int[n];
+        for (int[] arc : arcs) {
+            graph[arc[0]][currIdx[arc[0]]][0] = arc[1];
+            graph[arc[0]][currIdx[arc[0]]++][1] = arc[2];
+        }
+        return graph;
+    }
+
+    int dijkstra(int[][][] graph, int src, int dst, final int K, int n) {
+        int[] dist = new int[n];
+        int[] currStops = new int[n];
+        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
+        initializeDijkstra(src, n, dist, currStops, pq);
+
+        while (!pq.isEmpty()) {
+            int[] pathInfo = pq.poll();
+            int vertex = pathInfo[0];
+            int cost = pathInfo[1];
+            int stops = pathInfo[2];
+
+            if (vertex == dst) {
+                return cost;
+            }
+            if (stops == K + 1) {
+                continue;
+            }
+
+            for (int[] arc : graph[vertex]) {
+                int nextVertex = arc[0];
+                int altDist = cost + arc[1];
+
+                if (altDist < dist[nextVertex]) {
+                    pq.offer(new int[] { nextVertex, altDist, stops + 1 });
+                    dist[nextVertex] = altDist;
+                } else if (stops < currStops[nextVertex]) {
+                    pq.offer(new int[] { nextVertex, altDist, stops + 1 });
+                }
+                currStops[nextVertex] = stops;
```

Comment:
I wasn't able to rigorously prove why this is efficient. It might actually even be a heuristic. I will let it stay for now.

Reply:
As far as I understand; it gives a new consideration chance to an already visited node; if the node has a lesser stop value than it had before. 

 The reason is that the total distance to the node from the source might have been changed with the new configuration of the state.

  (By the way, It adds the same node multiple times into the PQ rather than updating the existing values of the node. Because otherwise, we would pay an extra search cost to find the node in the PQ.)

Reply:
Then let's think of some ways the next vertex can be handled, when it's going to be traversed to from some neighbor nodes:
```
1. Path: distance = 9, stops = 6
2. Path: distance = 13, stops 4
3. Path: distance = 11, stops 5
4. Path: distance = 20, stops 4
```

Of these 4 paths, only the 3. would not be added. But what if the destination was after this node and k equaled to 7? Then the most efficient path would have been the third one, which we did not add to the priority queue.

Reply:
I think we need to think in terms of nodes and edges instead of the distances in the first place.
Could you draw the graph according to these distances and stops you mentioned, please?

Reply:
![20210722_185214.jpg](https://user-images.githubusercontent.com/63192680/126669595-2825a8e2-de96-484a-b44d-0bd3709a2fcb.jpg)
I meant something like this.

Reply:
Thank you, Erdem. I will have a loot at it now.

Reply:
Actually, the node with length 9 will be polled out before the previous path to it with length 10. But I can find a new example that doesn't have such a property.

Reply:
![20210722_190122.jpg](https://user-images.githubusercontent.com/63192680/126670913-6ad23eda-3cb2-4212-a3d7-086392ac4147.jpg)
As in here.

Reply:
Then our test-case would be this, right?
```js
17
[[0,1,1],[0,2,1],[0,3,2],[0,4,3], [1,5,2], [5,6,1],[6,7,2], [7,8,1], [8,11,12], [2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4], [14,11,11], [4,15,6], [15,16,7], [16,11,11]]
0
11
1
```
![yes](https://user-images.githubusercontent.com/60903744/126674034-247cd7d7-9f2d-4932-9a2d-44c23313944a.jpg)

Reply:
> Then our test-case would be this, right?
> ```js
> 17
> [[0,1,1],[0,2,1],[0,3,2],[0,4,3], [1,5,2], [5,6,1],[6,7,2], [7,8,1], [8,11,12], [2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4], [14,11,11], [4,15,6], [15,16,7], [16,11,11]]
> 0
> 11
> 1
> ```
> ![yes](https://user-images.githubusercontent.com/60903744/126674034-247cd7d7-9f2d-4932-9a2d-44c23313944a.jpg)
> 

Destination is the node below it, with an arc to it of weight 100.

Reply:
Ah, oke, I fix it like this. Could you check, please?
```js
18
[[0,1,1],[0,2,1],[0,3,2],[0,4,3], [1,5,2], [5,6,1],[6,7,2], [7,8,1], [8,11,12], [2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4], [14,11,11], [4,15,6], [15,16,7], [16,11,11], [11,12,100]]
0
12
7
```

Reply:
```
18
[[0,1,1],[0,2,1],[0,3,2],[0,4,3], [1,5,2], [5,6,1],[6,7,2], [7,8,1], [8,11,12], [2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4], [14,11,11], [4,15,6], [15,16,7], [16,11,11], [11,12,100]]
0
12
6
```
It should have 6 stops. When I run this case, LeetCode outputs 3 though.

Reply:
I think this is the correct case:
```java
18
[[0,1,1],[0,2,1],[0,3,2],[0,4,6],[1,5,2], [5,6,1],[6,7,2], [7,8,1], [8,11,12],[2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4], [14,11,11], [4,15,6], [15,16,7], [16,11,11], [11,12,100]]
0
12
7
```

But, then again the output is 3.

Reply:
For
```
18
[[0,1,1],[0,2,1],[0,3,2],[0,4,3],[1,5,2],[5,6,1],[6,7,2],[7,8,1],[8,11,12],[2,9,4],[9,10,4],[10,11,14],[3,12,1],[12,13,3],[13,14,4],[14,11,11],[4,15,6],[15,16,7],[16,11,11],[11,17,100]]
0
17
5
```
It gave the wrong answer. I think we can conclude at this point that this line of code is just an imperfect heuristic, that somehow works through all the test cases of LeetCode. Dijkstra wasn't appropriate for this problem anyway.

Reply:
Oke, then I will just approve. The others look fine. When we turn back, we can have a look again.